// otplib/totp.js
/**
 * otplib-totp
 *
 * @author Gerald Yeo <contact@fusedthought.com>
 * @version: 10.1.0-0
 * @license: MIT
 **/
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var otplibCore = require('./core');
var hotp = _interopDefault(require('./hotp'));

var HOTP = hotp.HOTP;

var TOTP = function (_HOTP) {
  _inherits(TOTP, _HOTP);

  function TOTP() {
    _classCallCheck(this, TOTP);

    return _possibleConstructorReturn(this, (TOTP.__proto__ || Object.getPrototypeOf(TOTP)).call(this));
  }

  _createClass(TOTP, [{
    key: 'getClass',
    value: function getClass() {
      return TOTP;
    }
  }, {
    key: 'generate',
    value: function generate(secret) {
      var opt = this.optionsAll;
      return otplibCore.totpToken(secret || opt.secret, opt);
    }
  }, {
    key: 'check',
    value: function check(token, secret) {
      var delta = this.checkDelta(token, secret);
      return Number.isInteger(delta);
    }
  }, {
    key: 'checkDelta',
    value: function checkDelta(token, secret) {
      var opt = this.optionsAll;
      return otplibCore.totpCheckWithWindow(token, secret || opt.secret, opt);
    }
  }, {
    key: 'verify',
    value: function verify(opts) {
      if (typeof opts !== 'object' || opts == null) {
        return false;
      }
      return this.check(opts.token, opts.secret);
    }
  }, {
    key: 'timeRemaining',
    value: function timeRemaining() {
      var opt = this.optionsAll;
      return otplibCore.totpTimeRemaining(opt.epoch, opt.step);
    }
  }, {
    key: 'timeUsed',
    value: function timeUsed() {
      var opt = this.optionsAll;
      return otplibCore.totpTimeUsed(opt.epoch, opt.step);
    }
  }, {
    key: 'defaultOptions',
    get: function get() {
      return {
        epoch: null,
        step: 30,
        window: 0
      };
    }
  }, {
    key: 'optionsAll',
    get: function get() {
      return otplibCore.totpOptions(this._options);
    }
  }]);

  return TOTP;
}(HOTP);

TOTP.prototype.TOTP = TOTP;

var index = new TOTP();

module.exports = index;
